### PR TITLE
Add a nested level when selecting templates or template parts

### PR DESCRIPTION
**Pull request by youknowriad**
Related to #36667

## What?

This is a preparation PR for things like https://github.com/WordPress/gutenberg/issues/44461#issuecomment-1418950201 Basically in this PR we change the behavior of the site editor a bit to have a close relationship between what the sidebar shows and what's shown in the frame. 

So when you click a "template" or "a template part" you navigate to a page where you have the target template in the frame and the sidebar also updates to show details about the template.

## Notes

I've implemented this by making the "path" of the sidebar navigation as the "source of truth" from which all the rest (sidebar content and frame) adapts. The "path" is persited in the URL that way, you can come back to the same page.

There's an issue though: When you reload the page after selecting a template, the "back" link in the sidebar is broken, instead of navigating to the parent view (templates), it navigates to the root level one (design). I believe this is intrinsic to the design of the "Navigator" components and we need to adapt these in order to solve the issue. cc @ciampo 

This PR is not ready entirely though as there's still a lot of design work remaining in these new sidebar screens.

## Testing Instructions

1- Open the site editor
2- Navigate the site editor and select templates...

### CODE REVIEW

**Comment by jasmussen** [comment]
Nice work, this has much clearer and more contextual edit buttons:

status

I expect this will work well for Navigation as well.

We are missing descriptive text in the drilldown sidebar on the template drilldown, and when you press Back from a particular template, it'd be nice to go back to the main Templates list, rather than the top level. But other than that, this feels promising.

**Comment by jameskoster** [comment]
Nice! This feels pretty good to use :) 

Could we add a generic description for custom templates as a part of this PR? The panels for these templates look a bit empty. Perhaps something like:

> This is a custom template that can be applied manually to any Post or Page.

It would be good to come up with something for template parts too, let me know if that's something we can include here and I'll noodle on that.

In the future custom templates / template parts should be able to entertain custom descriptions but this should be fine to begin with.

---

Could we add descriptions to the Templates / Templates Parts / Navigation panels too?

---

If it's easy let's align the description with the site and panel title, and use $gray-600 for the color.

---

Could we include the chevron indicator on the menu items? Ideally the small chevron variant (which we should use at the root level too).

---

A couple of details that can be looked into separately:

* The blue edit button clashes a bit with the save button when it's visible. We may need to come up with a different variant.
* When there are many templates the 'Manage all templates' menu item can be hidden from view. 
* The small-screen experience on the 'Manage templates' view is really awkward.

**Comment by youknowriad** [comment]
> The small-screen experience on the 'Manage templates' view is really awkward.

I thought we could hide that menu in small screens entirely

**Comment by jameskoster** [comment]
Do you mean hide the management view? We might temporarily, but that would prohibit bulk actions on mobile (once we've added them) so it's not a long-term solution. These views will also need to be accessible if they're ever going to fulfill a role outside of the site editor too. 

I was thinking we might do something like this:



But it's certainly not something we need to handle in this PR.

**Comment by jameskoster** [comment]
I noticed couple of other issues: 

1. If you navigate into the Templates Panel, Select a template, click Edit, click W button, click the Back button then you get taken all the way back to the root instead of up one level to the Templates panel. 
2. Clicking a template in the management view doesn't do anything. Honestly I'm not sure what to do there, but since this view will likely become more about bulk editing it's probably okay. Still, we should update the styling so that the template names don't look like clickable links.

**Comment by youknowriad** [comment]
What kind of design tweaks we can do while we wait for the navigator updates to land?

**Comment by jasmussen** [comment]
Took it for a spin, it's working well. Here's a GIF going through the deeper nested drilldown menu:

site editor status

This is looking good. I don't think we necessarily need to block this PR from landing due to some of the small tweaks, but there are a couple of metrics we need to work out:
Screenshot 2023-02-10 at 13 13 08

I think we should use the 32px version of buttons for both the back button and the plus. I also wonder if we should show the plus below the list of templates, rather than fixed next to the title. With a black background and 24x24 footprint just like the in-canvas inserter.

The heading is a bit big. I have it as a semibold 16px in some mockups I've explored:

Screenshot 2023-02-10 at 13 16 11

I know @jameskoster has some thoughs on this too, so it might not be the right moment to finesse the details quite yet, we can probably do that after the fact.

Same with the deepest drilldown, we probably want to use the 32px button here:

Screenshot 2023-02-10 at 13 13 30

I find it a bit jarring that when I press "Back", I go to the top level instead of the higher level. I know there have been some component conversations here, but just for the record, here's what I'm seeing now:

* Starting in Design, I navigate to Templates → Frontpage. 
* When I press "Back" I would expect to go to "Templates", but I'm taken to "Design" again.

**Comment by youknowriad** [comment]
> I find it a bit jarring that when I press "Back", I go to the top level instead of the higher level. 

Yeah, this is the navigator updates I'm talking about this comment https://github.com/WordPress/gutenberg/pull/47777#issuecomment-1425613245 we're making progress in #47883. Once it lands, I'll rebase this PR and it should solve the issue.

For the button tweaks, they seem all kind of related, and hint to the introduction of a "new" button style in the sidebar, so maybe better to do in its own PR as well.

**Comment by jasmussen** [comment]
Cool, do you need a green check?

**Comment by youknowriad** [comment]
> Cool, do you need a green check?

It helps but I'll have to wait for #47883 anyway :)

**Review by jasmussen — APPROVED** [review]
This approach uxwise feels correct, and we'll explore polish separately.

**Comment by afercia** [comment]
> > I find it a bit jarring that when I press "Back", I go to the top level instead of the higher level.

> Yeah, this is the navigator updates I'm talking about this comment https://github.com/WordPress/gutenberg/pull/47777#issuecomment-1425613245 we're making progress in https://github.com/WordPress/gutenberg/pull/47883. Once it lands, I'll rebase this PR and it should solve the issue.

Cool, thanks 🙏  I was a bit concerned about that, as skipping a level would be unexpected for all users and even more confusing for users with accessibility needs.

Clarification for history: seems to me the skipping of the intermediate level occurs only when switching the editor to Edit mode. If I stay in browse mode, navigating back looks correct: Design ← Templates ← Front Page.

There's a problem with the button aria-label though.
When it skips a level and it goes from Front Page directly to Design, the aria-label still says:
`Navigate to the previous view: Templates`

Even if this gets fixed with #47883, I think the aria-label text is too complex. I do realize the good intent of providing more context but we need to find a compromise between non-sighted users and sighted users. For example, speech recognition software users would have a hard time figuring out the button accessible name is `Navigate to the previous view ...` while it visually looks like a 'Back' arrow. As a sighted speech recognition software user, I need to know the button accessible name to be able to issue a voice command. I'd consider to drastically simplify the aria-label and just use `Back`. This way, the voice command to issue would be the most intuitive one, based on the icon, e.g.: `Click Back`.
The contextual information `Navigate to the previous view: xyz` could go in a hidden element referenced by aria-describedby set on the button. I think I've seen this kind of hidden descriptions somewhere in the codebase (can't remember where off the top of my head).

**Comment by jameskoster** [comment]
Personally I'd like to see a description added for all panels that do not include menus.

Screenshot 2023-02-10 at 15 20 50

The screenshot above is sub-optimal, there's nothing in the panel to clarify what I'm looking at except the name which can be a bit arbitrary. I'm happy to provide descriptions, or we can add placeholders and refine while polishing other details.

The drilldown menu items are still missing a chevron. I suppose you can argue this is 'polish' but I'd say it has one foot in UX territory. 

The main outstanding issue: on the manage templates view clicking a template does nothing which feels like a pretty big regression, I'd consider this a blocker but happy to be overruled :) Imo we should either remove the link appearance, or make the links work (IE go directly to edit view).

**Comment by youknowriad** [comment]
@jameskoster I fixed the navigation from the "browse all" links.
@afercia I've simplified the back labels to just "Back"

File: packages/e2e-test-utils-playwright/src/site-editor/toggle-canvas-mode.js
Comment (kevin940726) [inline packages/e2e-test-utils-playwright/src/site-editor/toggle-canvas-mode.js]: Using CSS selectors is not recommended. Prefer [`getByRole`](https://playwright.dev/docs/locators#locate-by-role) instead.

```suggestion
	await this.page
		.getByRole( 'button', { name: 'Edit' } )
		.click();
```

(We might want to make this button more explicit with additional text or region for screen readers though.)

In addition, not something we need to do in this PR, but let's not create utils with only one or two lines if they can be [inlined](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/e2e/README.md#dont-overload-test-utils-inline-simple-utils).

Comment (youknowriad) [reply packages/e2e-test-utils-playwright/src/site-editor/toggle-canvas-mode.js]: Where does the "Edit" name come from, I don't think this has an "edit" name. It has an "Editor canvas" title instead. (It's an iframe that behaves like a button)

Comment (kevin940726) [reply packages/e2e-test-utils-playwright/src/site-editor/toggle-canvas-mode.js]: Isn't there still an "Edit" button in the sidebar? But if we want to target the iframe we can do:

```suggestion
	await this.page
		.getByRole( 'button', { name: 'Editor canvas' } )
		.click();
```

Comment (youknowriad) [reply packages/e2e-test-utils-playwright/src/site-editor/toggle-canvas-mode.js]: > Isn't there still an "Edit" button in the sidebar? 

No, not always, it's only present if your on some specific sidebars (single items).

Comment (youknowriad) [reply packages/e2e-test-utils-playwright/src/site-editor/toggle-canvas-mode.js]: It turns out neither of these work because we need to click inside the canvas instead. something like `canvas().click( 'body' )`

Comment (kevin940726) [reply packages/e2e-test-utils-playwright/src/site-editor/toggle-canvas-mode.js]: It seems to work if we specify the position explicitly:

```js
await page
	.getByRole( 'button', { name: 'Editor canvas' } )
	.click( { position: { x: 10, y: 10 } } );
```

By default, Playwright will find an arbitrary position, usually the center, to click on. But sometimes clicking on the center won't enter edit mode if the canvas is not full height. For instance, clicking on the bottom half of the canvas when viewing a small footer template part won't enter edit mode. This seems like a bug, is it?

(I guess specifying `{ x: 0, y: 0 }` also won't work because the canvas has `border-radius`.)

Comment (youknowriad) [reply packages/e2e-test-utils-playwright/src/site-editor/toggle-canvas-mode.js]: > By default, Playwright will find an arbitrary position, usually the center, to click on. But sometimes clicking on the center won't enter edit mode if the canvas is not full height. For instance, clicking on the bottom half of the canvas when viewing a small footer template part won't enter edit mode. This seems like a bug, is it?

Yeah, it's a bug I think but it's outside the scope of this PR. For now I removed the utility like you suggested and just went with canvas.

**Comment by ntsekouras** [comment]
@youknowriad you have made changes in php files for `6.3` but you've added the label to back port. Does this change needs to be in `6.2`?

**Comment by youknowriad** [comment]
@ntsekouras Good call, that was a mistake. It should target 6.2 instead.

**Comment by youknowriad** [comment]
@ntsekouras  I'll follow up with a PR that updates these files

**Comment by ntsekouras** [comment]
Cherry-picked this PR to the wp/6.2 branch.

**Comment by paaljoachim** [comment]
This is very helpful first step!
As it drills down into each Template and gives a description of each.

Here are some examples.
Screenshot 2023-02-15 at 06 44 29

Screenshot 2023-02-15 at 06 44 58

Screenshot 2023-02-15 at 06 45 10

All the templates contain a description.

**Template parts have no description. Should nesting be removed from Template parts because they do not contain any descriptions?**

**Comment by jameskoster** [comment]
> Template parts have no description. Should nesting be removed from Template parts because they do not contain any descriptions?

I don't think we should remove the nesting, it promotes consistency in terms of _how_ to edit (click the edit button) and creates space for us to add more features in the future. It would be good to add descriptions for these sections though.

**Comment by DomenicoCammarota** [comment]
tk

**Comment by afercia** [comment]
Noting that [the change to this background](https://github.com/WordPress/gutenberg/pull/47777/files#diff-6ffc115babb229ea68e4e9dadffa173abc901b2251e1c36d1322d067b6eedb0bR16), now applied also in 'view mode', broke the navigate region blue outline that is shown on focus:

Screenshot 2023-04-26 at 15 09 58

The logo container backgroudn now partially hides the blue outline. It happens also in 'edit mode', although I think that was already broken before this change.

Not the first time the navigate regions focus style breaks, and I'd really love to see some form of [visual regression testing](https://github.com/WordPress/gutenberg/pull/46210) in place into this project. WordCamp Europe would be a good opportunity to have some in person conversation about it.

In the meantime, I'll creat ea new issue.

**Pull request merged**
